### PR TITLE
Add support for go ecosystem metrics collection

### PR DESCRIPTION
**Pull request by amazimbe**
#### What are you trying to accomplish?

Update the go ecosystem to add support for gathering ecosystem metrics that need to be persisted in datadog via `POST /update_jobs/:id/record_ecosystem_meta`

#### Anything you want to highlight for special attention from reviewers?
This is one in a series of PRs to gather ecosystem meta data and persist it in datadog. See earlier PRs for [Bundler](https://github.com/dependabot/dependabot-core/pull/10826) and [NPM](https://github.com/dependabot/dependabot-core/pull/10862)

#### How will you know you've accomplished your goal?

- I'll be able to see Go ecosystem metrics in datadog

### Checklist

- [x] I have run the complete test suite to ensure all tests and linters pass.
- [x] I have thoroughly tested my code changes to ensure they work as expected, including adding additional tests for new functionality.
- [x] I have written clear and descriptive commit messages.
- [x] I have provided a detailed description of the changes in the pull request, including the problem it addresses, how it fixes the problem, and any relevant details about the implementation.
- [x] I have ensured that the code is well-documented and easy to understand.

### CODE REVIEW

File: go_modules/lib/dependabot/go_modules/file_parser.rb [hunk 1]
```diff
@@ -31,8 +33,40 @@ def parse
...
+          T.nilable(Dependabot::GoModules::PackageManager)
+        )
+      end
+
+      sig { returns(T.nilable(Ecosystem::VersionManager)) }
+      def language
+        @language ||= T.let(begin
+          Language.new(go_version)
+        end, T.nilable(Dependabot::GoModules::Language))
+      end
```
Comment (jakecoffman) [inline go_modules/lib/dependabot/go_modules/file_parser.rb:63]: These both appear to do the same thing, which doesn't seem very useful. 

Go doesn't have a separate package manager, it uses the toolchain version either from the `go` directive or the optional `toolchain` directive. Both of those would be useful to gather. You already have the first with `language` here, so I'd switch the `package_manager` to `toolchain` and figure out how to gather that, if present.

Comment (amazimbe) [reply go_modules/lib/dependabot/go_modules/file_parser.rb:63]: Ok, I'll take a look at how I can get the toolchain

Comment (amazimbe) [reply go_modules/lib/dependabot/go_modules/file_parser.rb]: This is done

File: go_modules/lib/dependabot/go_modules/package_manager.rb [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    class PackageManager < Dependabot::Ecosystem::VersionManager
+      extend T::Sig
+
+      sig do
+        params(
+          raw_version: String,
+          requirement: T.nilable(Requirement)
+        ).void
+      end
+      def initialize(raw_version, requirement = nil)
```
Comment (kbukum1) [inline go_modules/lib/dependabot/go_modules/package_manager.rb:27]: If for go you don't have requirements you don't need to add it to your initialize method?

Comment (amazimbe) [reply go_modules/lib/dependabot/go_modules/package_manager.rb]: I've cleaned this up.

**Review by kbukum1 — APPROVED** [review]
Really nice works!

**Comment by popchili** [comment]
Espero que esto se vuelva una o dos rutinas 322271724

**Pull request merged**